{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f5b58a9d",
   "metadata": {},
   "source": [
    "# One-file code for the simulations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2619c993",
   "metadata": {},
   "source": [
    "To install all required packages and dependencies, run the following cell (only needs to be run once):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0108840f",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Pkg\n",
    "Pkg.add(\"Random\")\n",
    "Pkg.add(\"Distributions\")\n",
    "Pkg.add(\"DataFrames\")\n",
    "Pkg.add(\"StatsBase\")\n",
    "Pkg.add(\"CairoMakie\")\n",
    "Pkg.add(\"ProgressMeter\")\n",
    "Pkg.add(\"JLD2\")\n",
    "Pkg.add(\"Integrals\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4d30221",
   "metadata": {},
   "source": [
    "## Model implementation code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb2ed65a",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Pkg\n",
    "using Random, Distributions, DataFrames, StatsBase, CairoMakie, ProgressMeter, JLD2, Integrals\n",
    "\n",
    "\n",
    "\n",
    "# Core model elements\n",
    "# Mating model\n",
    "\n",
    "Mating = function (; Nf, Nm, E, ff, r_sex, beta=0, M2=10, M=4, tp=Int64,wa=1)\n",
    "    # Pre-allocate arrays\n",
    "    N_pairs_A = zeros(tp, M2, M2)\n",
    "    N_pairs_Af = zeros(tp, M2, M2)\n",
    "    N_exp_f = zeros(tp, M2)\n",
    "    N_exp_m = zeros(tp, M2)\n",
    "    \n",
    "    Nm_sum = sum(Nm.*wa)\n",
    "    if Nm_sum == 0\n",
    "        return N_exp_f, N_exp_m  # Early return if no males\n",
    "    end\n",
    "    pAm = Nm.*wa ./ Nm_sum\n",
    "    p_beta = Nm_sum / (beta + Nm_sum)\n",
    "\n",
    "    # Vectorized mating pair sampling\n",
    "    @inbounds for i in 1:M2\n",
    "        if Nf[i] > 0\n",
    "            n_pairs = round(tp, Nf[i] * p_beta)\n",
    "            if n_pairs > 0\n",
    "                N_pairs_A[i, :] .= rand(Multinomial(n_pairs, pAm))\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # Vectorized offspring production\n",
    "    @inbounds for j in 1:M2, i in 1:M2\n",
    "        if N_pairs_A[i, j] > 0\n",
    "            # Sample fertile pairs and offspring in one step\n",
    "            N_pairs_Af[i, j] = rand(Poisson(N_pairs_A[i, j] * ff[i, j]))\n",
    "            \n",
    "            if N_pairs_Af[i, j] > 0\n",
    "                N_exp = rand(Multinomial(N_pairs_Af[i, j], E[i][j]))\n",
    "                # Sex allocation - vectorized\n",
    "                N_sx = rand.(Binomial.(N_exp, 1 .- r_sex[j, i]))\n",
    "                \n",
    "                N_exp_f .+= N_sx\n",
    "                N_exp_m .+= N_exp .- N_sx\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    return N_exp_f, N_exp_m\n",
    "end\n",
    "\n",
    "# Fitness change: survival/sterility effects in adults\n",
    "Fitness_change = function (; Nf, Nm, wf, wm, M2=10)\n",
    "    Nf_new = first.(rand.(Binomial.(Nf, wf), 1))\n",
    "    Nm_new = first.(rand.(Binomial.(Nm, wm), 1))\n",
    "    return Nf_new, Nm_new\n",
    "end\n",
    "#Density-dependent mortality\n",
    "Alpha=function(;N, f,phi1=1, phi2=1,tp=Int64)\n",
    "    if phi2==1\n",
    "        a = (N*f/2)/(phi1*f/2-1)\n",
    "    else \n",
    "        a= (N*f/2)/(phi1*f/2-1)^(1/phi2)\n",
    "    end\n",
    "    return a\n",
    "end\n",
    "\n",
    " Ddm_change=function(; Nf, Nm, Nh, alpha, M2=10, tp=Int64,phi1=1,phi2=1)\n",
    "    p = phi1 / ((Nh/alpha)^phi2 + 1)\n",
    "    \n",
    "    # Pre-allocate and fill directly\n",
    "    Nf_new = Vector{tp}(undef, M2)\n",
    "    Nm_new = Vector{tp}(undef, M2)\n",
    "    \n",
    "    @inbounds for i in 1:M2\n",
    "        Nf_new[i] = rand(Binomial(Nf[i], p))\n",
    "        Nm_new[i] = rand(Binomial(Nm[i], p))\n",
    "    end\n",
    "    \n",
    "    return Nf_new, Nm_new\n",
    "end\n",
    "\n",
    "# Model for 1 generational change\n",
    "# Ntfi/Ntmi for total adult population, Ntf/Ntm for fertile adult population, counted per genotype\n",
    "Model2 = function (; Nf, Nm, E, wm, wf, alpha=10^6, ff, r_sex=0.5, beta=1, M2=10, M=4,tp=Int64,phi1=1,phi2=1,wa=1)\n",
    "    Nf, Nm = Mating(; Nf=Nf, Nm=Nm, E=E, ff=ff, r_sex=r_sex, beta=beta, M2=M2, M=M,tp=tp,wa=wa)\n",
    "    Nh = sum(Nf .+ Nm)\n",
    "    Nf_1, Nm_1 = Ddm_change(; Nf=Nf, Nm=Nm, Nh=Nh, alpha=alpha, M2=M2,phi1=phi1,phi2=phi2)[1:2]\n",
    "    Nf_2, Nm_2 = Fitness_change(; Nf=Nf_1, Nm=Nm_1, wf=wf, wm=wm, M2=M2)\n",
    "    return [Nf_2, Nm_2]#Ntf, Ntm, Ntfi, Ntmi\n",
    "end\n",
    "\n",
    "# Population size input\n",
    "Input_nosp = function (; M=4, Nmean=10^5, pm=0.01,tp=Int64)\n",
    "    M2 = Int(M * (M + 1) / 2)\n",
    "    Ntf_init = zeros(tp,M2)\n",
    "    Ntf_init[1] = tp(round(Nmean / 2))\n",
    "    Ntm_init = zeros(tp,M2)\n",
    "    Ntm_init[1] = tp(round(Nmean / 2))\n",
    "    Ntm_init[M+1] = ceil(tp, Ntm_init[1] * pm)\n",
    "    Ntm_init[1] = ceil(tp, Ntm_init[1] * (1 - pm))\n",
    "    return Ntf_init, Ntm_init\n",
    "end\n",
    "\n",
    "# Full simulations over time. Tracking only cumulative population size by sex\n",
    "# Parameters: maxgen=number of generations\n",
    "# N=total effective population size\n",
    "# beta=parameter of probability of successfully finding a male mate in small population, e.g., p=(number of males)/(number of males+beta). Not used in non-spatial modelling\n",
    "# wmf=fitness and reproduction data, generated by Fitness_input1 function\n",
    "# e=gametes production data,  generated by E_Make_1 function\n",
    "# alpha=parameter related to density depending mortality. Alpha=0 for equilibrium = N\n",
    "# phi1, phi2 are parameters from the density-dependency model (Maynard Smith & Blatkin). When phi1=phi2=1 is equal to the Beverton-Holt model\n",
    "# M=4 or M=5 = number of alleles (M=5 for drive with no sex distorter)\n",
    "#pm = initial DD drive males release frequency\n",
    "\n",
    " Model_nosp_NT=function(; N=10^15, pm=0.01, wmf,e, mD_sex, beta=0, alpha=0,phi1=1,phi2=1, maxgen=10, Model=2, M=4,tp=Int64,b2=0)\n",
    "    Ntf_init, Ntm_init = Input_nosp(; M=M, Nmean=N, pm=pm,tp=tp)\n",
    "   \n",
    "    M2 = Int(M * (M + 1) / 2)\n",
    "    f=wmf[3][1,1]\n",
    "    alpha = alpha == 0 ? Alpha(phi1=phi1,phi2=phi2, f=f,N=N) : alpha\n",
    "\n",
    "    r_sex = 0.5 .* ones(M2, M2)\n",
    "    if M == 4\n",
    "        for i in [2, 5, 6, 7]\n",
    "            r_sex[i, :] .= mD_sex\n",
    "        end\n",
    "    end\n",
    "    if M == 5\n",
    "        for i in [2, 6, 7, 8, 9]\n",
    "            r_sex[i, :] .= mD_sex\n",
    "        end\n",
    "    end\n",
    "    if b2>0     \n",
    "        if M==4 r_sex[7,:] .=0.5 ;end\n",
    "        if M==5 r_sex[9,:] .=0.5 ;end\n",
    "    end\n",
    "    model = Model == 2 ? Model2 : Model1\n",
    "\n",
    "    NCountF=Array{tp}(undef, maxgen)\n",
    "    NCountM=Array{tp}(undef, maxgen)\n",
    "\n",
    "    \n",
    "    Ntf, Ntm = copy(Ntf_init), copy(Ntm_init)\n",
    "    NCountF[1]=sum(Ntf_init)\n",
    "    NCountM[1]=sum(Ntm_init)\n",
    "\n",
    "    for gen in 1:(maxgen - 1)\n",
    "        # Reproduction change\n",
    "       \n",
    "        if (NCountM[gen]>1) * (NCountF[gen]>1)> 0\n",
    "            beta = NCountM[gen] / (NCountM[gen] + beta)\n",
    "            m = model(; Nf=Ntf, Nm=Ntm, E=e, alpha=alpha, ff=wmf[3], wm=wmf[1], wf=wmf[2], r_sex=r_sex,phi1=phi1,phi2=phi2, beta=beta, M2=M2, M=M,tp=tp)\n",
    "            Ntf = m[1]\n",
    "            Ntm = m[2]\n",
    "            \n",
    "        else\n",
    "            Ntf, Ntm = zeros(tp, M2), zeros(tp, M2)\n",
    "        end\n",
    "        NCountM[gen+1] = sum(Ntm)\n",
    "        NCountF[gen+1] = sum(Ntf)\n",
    "    end\n",
    "    dc_ = [NCountF, NCountM]\n",
    "    return dc_\n",
    "end\n",
    "\n",
    "\n",
    "#Additional functions\n",
    "Genot=function(;A,M2,M)\n",
    "    v = Array{Float64}(undef,M2)\n",
    "    k=0\n",
    "    for i in 1:M\n",
    "        v[k+=1] =A[i,i]\n",
    "        for j in (i+1):M\n",
    "            v[k+=1]=A[i,j]+A[j,i]\n",
    "        end\n",
    "    end\n",
    "    return v\n",
    "end\n",
    "\n",
    "# Gametes produced probability (genotype) \n",
    "E_Make_1=function(;c=0.95,j=0.03,a=0.01,b=0.001,show=false,M=4,title=\"Gametes produces\")\n",
    "    if M==5\n",
    "        E5=vcat(\n",
    "            [1 0 0 0 0],\n",
    "            [(1-c)/2 (1/2+c*(1-j)*(1-a)/2) c*j*(1-b)/2 c*j*b/2 c*(1-j)*a/2],\n",
    "            [1/2 0 1/2 0 0],\n",
    "            [1/2 0 0 1/2 0],\n",
    "            [(1-c)/2 0 c*j*(1-b)/2 c*j*b/2 (1/2+c*(1-j)/2)],\n",
    "            [0 1 0 0 0],\n",
    "            [0 1/2 1/2 0 0],\n",
    "            [0 1/2 0 1/2 0],\n",
    "            [0 1/2 0 0 1/2],\n",
    "            [0 0 1 0 0],\n",
    "            [0 0 1/2 1/2 0],\n",
    "            [0 0 1/2 0 1/2],\n",
    "            [0 0 0 1 0],\n",
    "            [0 0 0 1/2 1/2],\n",
    "            [0 0 0 0 1]\n",
    "            )\n",
    "\n",
    "    end\n",
    "    if M==4\n",
    "        E5=vcat(\n",
    "            [1 0 0 0 ],\n",
    "            [(1-c)/2 (1/2+c*(1-j)/2) c*j*(1-b)/2 c*j*b/2 ],\n",
    "            [1/2 0 1/2 0 ],\n",
    "            [1/2 0 0 1/2 ],\n",
    "            [0 1 0 0 ],\n",
    "            [0 1/2 1/2 0 ],\n",
    "            [0 1/2 0 1/2 ],\n",
    "            [0 0 1 0 ],\n",
    "            [0 0 1/2 1/2 ],\n",
    "            [0 0 0 1 ]\n",
    "            )\n",
    "\n",
    "    end    \n",
    "    return transpose(E5)\n",
    "end\n",
    "# Gametes produced probability (sex-specific homing)\n",
    "\n",
    "# f x m gametes pairs produced by each genotype pair\n",
    "E_mat=function(;Ef,Em,M2,M)\n",
    "    E_temp=[[Ef[:,i]*transpose(Em[:,j]) for i in 1:M2] for j in 1:M2]\n",
    "    e=[[Genot(;A=E_temp[i][j],M=M,M2=M2) for i in 1:M2] for j in 1:M2]\n",
    "    return e\n",
    "end\n",
    "\n",
    "# gametes production final function. Allows for homing in female/males/both\n",
    "E_Make = function (; M=4, homing=:b, c_m=0.9, j_m=035, a_m=0, b_m=0.001, c_f=0.9, j_f=0.035, a_f=0, b_f=0.001, show=false)\n",
    "    M2 =Int(M*(M+1)/2)\n",
    "    if homing == :b\n",
    "        if show\n",
    "            println(\"Homing in both sexes\")\n",
    "        end\n",
    "        Em = E_Make_1(; c=c_m, j=j_m, a=a_m, b=b_m, show=show, M=M, title=\"Gametes produced (males)\")\n",
    "        Ef = E_Make_1(; c=c_f, j=j_f, a=a_f, b=b_f, show=show, M=M, title=\"Gametes produced (females)\")\n",
    "    elseif homing == :m\n",
    "        if show\n",
    "            println(\"Homing in males\")\n",
    "        end\n",
    "        Em = E_Make_1(; c=c_m, j=j_m, a=a_m, b=b_m, show=show, M=M, title=\"Gametes produced (males)\")\n",
    "        Ef = E_Make_1(; c=0, j=0, a=0, b=0, show=false, M=M, title=\"Gametes produced (females)\")\n",
    "    elseif homing == :f\n",
    "        if show\n",
    "            println(\"Homing in females\")\n",
    "        end\n",
    "        Em = E_Make_1(; c=0, j=0, a=0, b=0, show=false, M=M, title=\"Gametes produced (males)\")\n",
    "        Ef = E_Make_1(; c=c_f, j=j_f, a=a_f, b=b_f, show=show, M=M, title=\"Gametes produced (females)\")\n",
    "    end\n",
    "    e = E_mat(; Ef=Ef, Em=Em, M2=M2, M=M)\n",
    "    return e\n",
    "end\n",
    "\n",
    "Fitness_input1=function(;sigma=0.02,s=1,h=1,hN=0.03,hDR=0.03,show=true,M=4,hm=0,f=12,inter=true,wf_nb=[],non_func=[\"D\",\"N\",\"O\"])\n",
    "    M2=Int(M*(M+1)/2)\n",
    "    wm=ones(M2)\n",
    "    if M==5\n",
    "        wf0=[1,\n",
    "        1-h*s,\n",
    "        1-hN*s,\n",
    "        1-sigma,\n",
    "        1-h*s,\n",
    "        1-s,\n",
    "        1-s,\n",
    "        (1-sigma)*(1-hDR*s),\n",
    "        1-s,\n",
    "        1-s,\n",
    "        (1-sigma)*(1-hN*s),\n",
    "        1-s,\n",
    "        (1-sigma)^2,\n",
    "        (1-sigma)*(1-hDR*s),\n",
    "        1-s\n",
    "        ]\n",
    "        wfth=[\"1\",\n",
    "        \"1-hs\",\n",
    "        \"1-hN*s\",\n",
    "        \"1-sigma\",\n",
    "        \"1-h*s\",\n",
    "        \"1-s\",\n",
    "        \"1-s\",\n",
    "        \"(1- sigma)(1-hDRs)\",\n",
    "        \"1-s\",\n",
    "        \"1-s\",\n",
    "        \"(1-sigma)(1-hN*s)\",\n",
    "        \"1-s\",\n",
    "        \"(1-sigma)^2\",\n",
    "        \"(1-sigma)(1-hDR*s)\",\n",
    "        \"1-s\"\n",
    "        ]\n",
    "        namesG=[\"WW\",\"WD\",\"WN\",\"WR\",\"WO\",\"DD\",\"DN\",\"DR\",\"DO\",\"NN\",\"NR\",\"NO\", \"RR\",\"RO\",\"OO\"]\n",
    "        \n",
    "     end\n",
    "    if M==4\n",
    "        wf0=[1,\n",
    "        1-h*s,\n",
    "        1-hN*s,\n",
    "        1-sigma,\n",
    "        1-s,\n",
    "        1-s,\n",
    "        (1-sigma)*(1-hDR*s),\n",
    "        1-s,\n",
    "        (1-sigma)*(1-hN*s),\n",
    "        (1-sigma)^2,\n",
    "        ]\n",
    "        wfth=[\"1\",\n",
    "        \"1-h*s\",\n",
    "        \"1-h_N*s\",\n",
    "        \"1-sigma\",\n",
    "        \"1-s\",\n",
    "        \"1-s\",\n",
    "        \"(1-sigma)(1-hDR*s)\",\n",
    "        \"1-s\",\n",
    "        \"(1-sigma)(1-hN*sN)\",\n",
    "        \"(1-sigma)^2\",\n",
    "        ]\n",
    "        namesG=[\"WW\",\"WD\",\"WN\",\"WR\",\"DD\",\"DN\",\"DR\",\"NN\",\"NR\",\"RR\"]\n",
    "                \n",
    "    end\n",
    "    namesF=[\"Males\",\"Females\",\"Formula (female)\"]\n",
    "   if inter\n",
    "        wf=ones(M2)\n",
    "        wf[wf0.<=0] .=0\n",
    "    else wf=wf_nb\n",
    "    end\n",
    "    di=occursin.(\"D\",namesG);wm[di].=1-hm\n",
    "    \n",
    "    ff=ones(M2,M2).*f\n",
    "    for j in 1:M2 for i in 1:M2 ff[i,j]=wf0[i]*wm[j]*f;end;end\n",
    "    if show \n",
    "        pretty_table([wm wf0 wfth ],header=namesF,tf=tf_borderless,row_labels=namesG,alignment=:l)\n",
    "    end\n",
    "    return wm,wf,ff\n",
    "end\n",
    "\n",
    "#Integrated suppression area calculation for a population count output\n",
    "IntegralCum=function(;maxgen=400,dat,n)\n",
    "    method = TrapezoidalRule()\n",
    "    arr=[]\n",
    "    for t in 2:maxgen\n",
    "        problem=SampledIntegralProblem((n/2 .-dat[1:t])./(n/2), 1:t)\n",
    "        push!(arr,solve(problem, method).u)\n",
    "    end\n",
    "    return arr\n",
    "end\n",
    "\n",
    "#Duration of protection calculation based on population count data. Returns number of generations when population is reduced to <=p pf its original size\n",
    "TimePerc=function(n;maxt=300,perc=0.5)\n",
    "    n0=n[1]\n",
    "    tt=filter(x->n[x]<=n0*perc,1:maxt)\n",
    "    if length(tt)==0\n",
    "        tt=[1,1]\n",
    "    end\n",
    "    return first(tt),last(tt),last(tt)-first(tt)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d207a3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "include(\"SDMDModel.jl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e7ef70",
   "metadata": {},
   "outputs": [],
   "source": [
    "Random.seed!(1234);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "967e0e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "using CairoMakie\n",
    "set_theme!(theme_minimal()) \n",
    "\n",
    "colSDMD=(cgrad(:winter, categorical=true,3)[1],0.6)\n",
    "cols3=[:black,:grey,colSDMD]\n",
    "\n",
    "labelt= [\"GD\",\"SDGD\",\"SDMD\"]\n",
    "marks=[:star5,:cross,:circle] \n",
    "lnst=[:solid,:solid,:solid];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff6c335",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running simulations for 3 strategies\n",
    "#r1 - sex distortion, Nvec = vector of the population sizes\n",
    "\n",
    "Runs_Nosp_dx=function(;N,r1,rep=500,C=0.9,B=0.001,J=0.05,F=12,hf,hDR=0,hN=0,hm,A=0,maxgen=400,Release=0.01,filename1=\"dx\",homing=:b,beta=0,M=4,Model=2,inter=true,wf_nb=[],dir=\"\")\n",
    "    N_temp1=[]    #full model output\n",
    "    N_tempF=[]    #female population (fertile)\n",
    "    N_tempI=[]    #itegrated suppression area \n",
    "\n",
    "    M2=Int(M*(M+1)/2)\n",
    "    e = E_Make(;c_m=C,c_f=C,j_m=J,j_f=J,b_m=B,b_f=B,a_f=A,a_m=A,M=M,homing=:b)\n",
    "    wmf=Fitness_input1(;sigma=0,s=1,h=hf,hN=hN,hDR=hDR,hm=hm,show=false,M=M,inter=inter, wf_nb=wf_nb)\n",
    "\n",
    "    fn1=string(dir,filename1,\"_m\",r1,\"_hf\",hf,\"_hN\",hN,\"_hm\",hm,\"_A\",A,\"_C\",C,\"_B\",B,\"_J\",J,\".jld2\")\n",
    "    fn2=string(dir,\"F\",filename1,\"_m\",r1,\"_hf\",hf,\"_hN\",hN,\"_hm\",hm,\"_A\",A,\"_C\",C,\"_B\",B,\"_J\",J,\".jld2\")\n",
    "    fn3=string(dir,\"Int\",filename1,\"_m\",r1,\"_hf\",hf,\"_hN\",hN,\"_hm\",hm,\"_A\",A,\"_C\",C,\"_B\",B,\"_J\",J,\".jld2\")\n",
    "    @showprogress for rp in 1:rep\n",
    "        sleep(0.1)\n",
    "\n",
    "        mod1=Model_nosp_NT(;N=N,wmf=wmf,e=e,mD_sex=r1,M=M,maxgen=maxgen,beta=beta,pm=Release)\n",
    "        n=mod1[1]\n",
    "        push!(N_tempI,IntegralCum(;dat=n,n=n[1]*2))\n",
    "        push!(N_tempF,n)\n",
    "        push!(N_temp1,mod1)\n",
    "    end\n",
    "    jldsave(fn1, true; models1=N_temp1)   #save outputs\n",
    "    jldsave(fn2, true; models1=N_tempF) \n",
    "    jldsave(fn3, true; models1=N_tempI) \n",
    "    return N_tempF,N_tempI\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f4ccdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "IdealisedRuns=[\n",
    "    Runs_Nosp_dx(;N=N0,r1=0.5,rep=rep,C=0.95,B=0.005,J=0.035,F=12,hf=0,hN=0,hm=0,A=0,filename1=\"dgd\"), #GD\n",
    "\n",
    "    Runs_Nosp_dx(;N=N0,r1=0.95,rep=rep,C=0.95,B=0.005,J=0.035,F=12,hf=0,hN=0,hm=0,A=0,filename1=\"dsdgd\") ,#SDGD\n",
    "\n",
    "    Runs_Nosp_dx(;N=N0,r1=0.95,rep=rep,C=0.95,B=0.005,J=0.035,F=12,hf=1,hN=0,hm=0,A=0,inter=false, wf_nb=SDMD_wf,filename1=\"dx\") ,#SDMD \n",
    "\n",
    "];\n",
    "#Duration of protection \n",
    "p=0.5\n",
    "TRest_095= [mean([TimePerc(IdealisedRuns[h][1][i];perc=p)[2]-TimePerc(IdealisedRuns[h][1][i];perc=p)[1] for i in 1:rep]) for h in eachindex(IdealisedRuns)];\n",
    "#Duration of protection \n",
    "p=0.5\n",
    "TRest_095= [mean([TimePerc(IdealisedRuns[h][1][i];perc=p)[2]-TimePerc(IdealisedRuns[h][1][i];perc=p)[1] for i in 1:rep]) for h in eachindex(IdealisedRuns)];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cac04046",
   "metadata": {},
   "outputs": [],
   "source": [
    "#IMPACT OF THE FITNESS ON THE POPULATION DYNAMICS (SEX DISTORTION=0.9, 0.99)\n",
    "#b = 0.005, and HNEJ rates j= 0.035\n",
    "f=Figure(size=(900,400))\n",
    "#supertitle=Label(f[0,1:3], \"Sex distortion and W/D female fitness effects on population suppression\",fontsize=16,halign=:left)\n",
    "ax1=Axis(f[1,1],title=\"a\",titlealign=:left,ylabel=\"Relative female population size\",xlabel=\"Generations\")\n",
    "gen=100\n",
    "\n",
    "n=N0\n",
    "mods=first.(IdealisedRuns) #population\n",
    "modsI=last.(IdealisedRuns) #integrated area\n",
    "\n",
    "for i in 1:3\n",
    "    scatterlines!(ax1,1:gen,mean([mods[i][j][1:gen]./(0.5*n) for j in 1:rep]) ,color=cols3[i],marker=marks[i],markersize=6)\n",
    "    band!(ax1,1:gen,mean([mods[i][j][1:gen]./(0.5*n) for j in 1:rep]),repeat([1],gen),alpha=0.1,color=(cols3[i],0.1),label=labelt[i])\n",
    "\n",
    "end\n",
    "\n",
    "\n",
    "ax2=Axis(f[1,2],title=\"b  \",ylabel=\"Integrated suppression area\",xticks=(1:3,labelt),xticklabelrotation=pi/2,titlealign=:left,xticklabelsize=11)\n",
    "is=[modsI[i][1][end] for i in 1:3]\n",
    "barplot!(ax2,1:3,is,color=cols3,gap=0.1,strokecolor = :black, strokewidth = 0)#,direction=:x,color=col66)\n",
    "\n",
    "\n",
    "ax3=Axis(f[1,3],title=\"c  \",ylabel=\"Duration of 50% suppression\",xticks=(1:3,labelt),xticklabelrotation=pi/2,titlealign=:left,xticklabelsize=11)\n",
    "ts=[TRest_095[i] for i in 1:3]\n",
    "barplot!(ax3,1:3,ts,color=cols3,gap=0.1,strokecolor = :black, strokewidth = 0.5)\n",
    "\n",
    "leg=[MarkerElement(color = cols3[i],markersize=13,marker=marks[i]) for i in 1:3];\n",
    "Legend(f[0, 1:3], leg, [\"GD\",\"SDGD\",\"SDMD\"],\"\",titleposition=:left,orientation = :horizontal,padding=(0,0,0,0),halign=:left,colgap =15,tellwodth=false)#,framecolor=:lightgrey,framevisible=true)\n",
    "Box(f[0, 1:3],color=(:white,0),strokewidth = 1,strokecolor=:lightgrey)\n",
    "\n",
    "colsize!(f.layout, 1, Relative(3.6/6))\n",
    "colsize!(f.layout, 2, Relative(1.2/6))\n",
    "colsize!(f.layout, 3, Relative(1.2/6))\n",
    "save(\"SDMDFigure1_idealised.png\",f)\n",
    "f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8115cd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "b=0.0016\n",
    "EmpiricalRuns=\n",
    "    [Runs_Nosp_dx(;N=N0,r1=0.5,rep=rep,C=c_GD,B=b,J=j_GD,F=12,hf=0,hN=0,hm=0,A=0,filename1=\"gdEmpB\"),\n",
    "     Runs_Nosp_dx(;N=N0,r1=r_SDGD,rep=rep,C=c_SDGD,B=b,J=j_SDGD,F=12,hf=0.38,hN=0,hm=0,A=0,filename1=\"sdgdEmpB\"),\n",
    "     Runs_Nosp_dx(;N=N0,r1=r_SDMD,rep=rep,C=c_SDMD,B=b,J=j_SDMD,F=12,hf=1,hDR=1,hN=0,hm=0,A=0,filename1=\"dxEmpB\")\n",
    "    ];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f851cc74",
   "metadata": {},
   "outputs": [],
   "source": [
    "p=0.5\n",
    "#DYNAMICS, INTEGRATED SUPPRESSION AREA AND DURATION OF PROTECTION FOR 3 EMPIRICAL CONSTRUCTS\n",
    "f=Figure(size=(950,400))\n",
    "\n",
    "n=N0\n",
    "gen=150\n",
    "mods=first.(EmpiricalRuns)\n",
    "modsI=last.(EmpiricalRuns)\n",
    "\n",
    "ax1=Axis(f[1,1],title=\"a \", ylabel=\"Relative female population size\",xlabel=\"Generations\",titlealign=:left)\n",
    "for i in 1:3\n",
    "    scatterlines!(ax1,1:gen,mean([mods[i][j][1:gen]./(0.5*n) for j in 1:rep]) ,color=cols3[i],marker=marks[i],markersize=6)\n",
    "    band!(ax1,1:gen,mean([mods[i][j][1:gen]./(0.5*n) for j in 1:rep]),repeat([1],gen),alpha=0.1,color=(cols3[i],0.1),label=labelt[i])\n",
    "\n",
    "end\n",
    "\n",
    "ax2=Axis(f[1,2],title=\"b  \",ylabel=\"Integrated suppression area\",xticks=(1:3,[\"GD\",\"SDGD\",\"SDMD\"]),xticklabelrotation=pi/2,titlealign=:left)\n",
    "\n",
    "y=[mean([modsI[i][j][end] for j in 1:rep]) for i in 1:3]\n",
    "\n",
    "barplot!(ax2,1:3, y,color=cols3,strokecolor = :black, strokewidth = 0.5)\n",
    "ax3=Axis(f[1,3],title=\"c  \",ylabel=\"Duration of 50% suppression\",xticks=(1:3,[\"GD\",\"SDGD\",\"SDMD\"]),titlealign=:left,xticklabelrotation=pi/2)\n",
    "TRestEmp= [TimePerc(mods[i][1];perc=p)[2]-TimePerc(mods[i][1];perc=p)[1] for i in 1:3];\n",
    "barplot!(ax3,1:3,TRestEmp[1:3],color=cols3,strokecolor = :black, strokewidth = 0.5)#,direction=:x)\n",
    "\n",
    "colsize!(f.layout, 1, Relative(4/6))\n",
    "colsize!(f.layout, 2, Relative(1/6))\n",
    "colsize!(f.layout, 2, Relative(1/6))\n",
    "\n",
    "leg=[MarkerElement(color = cols3[i],markersize=13,marker=marks[i]) for i in 1:3];\n",
    "Legend(f[0, 1:3], leg, [\"GD\",\"SDGD\",\"SDMD\"],\"\",titleposition=:left,orientation = :horizontal,padding=(0,0,0,0),halign=:left,colgap =15,tellwodth=false)#,framecolor=:lightgrey,framevisible=true)\n",
    "Box(f[0, 1:3],color=(:white,0),strokewidth = 1,strokecolor=:lightgrey)\n",
    "\n",
    "save(\"SDMDFigure2_empirical.png\",f)\n",
    "f"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.1",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
